core/src/Cabana_Parallel.hpp

```diff
@@ -554,6 +576,422 @@ inline void neighbor_parallel_for(
+                ival );
+    };
+    if ( str.empty() )
+        Kokkos::parallel_reduce( exec_policy, neigh_reduce, reduce_val );
+    else
+        Kokkos::parallel_reduce( str, exec_policy, neigh_reduce, reduce_val );
+}
+
+//---------------------------------------------------------------------------//
+/*!
+  \brief Execute \c functor reduction in parallel according to the execution \c
+  policy with thread-local serial loops over particle first and second
+  neighbors.
+
+  \tparam FunctorType The functor type to execute.
+
+  \tparam NeighborListType The neighbor list type.
+
+  \tparam ExecParams The Kokkos range policy parameters.
+
+  \tparam ReduceType The reduction type.
+
+  \param exec_policy The policy over which to execute the functor.
+
+  \param functor The functor to execute in parallel
+
+  \param list The neighbor list over which to execute the neighbor operations.
+
+  \param neighborstag Iteration tag indicating operations over particle first
+  and second neighbors.
+
+  \param optag Algorithm tag indicating a serial loop strategy over particle
+  neighbors.
+
+  \param reduce_val Scalar to be reduced across particles and neighbors.
+
+  \param str Optional name for the functor. Will be forwarded if non-empty to
+  the Kokkos::parallel_reduce called by this code and can be used for
+  identification and profiling purposes.
+*/
+template <class FunctorType, class NeighborListType, class ReduceType,
+          class... ExecParameters>
+inline void neighbor_parallel_reduce(
+    const Kokkos::RangePolicy<ExecParameters...> &exec_policy,
+    const FunctorType &functor, const NeighborListType &list,
+    const SecondNeighborsTag, const SerialOpTag, ReduceType &reduce_val,
+    const std::string &str = "" )
+{
+    using work_tag = typename Kokkos::RangePolicy<ExecParameters...>::work_tag;
+
+    using execution_space =
+        typename Kokkos::RangePolicy<ExecParameters...>::execution_space;
+
+    using index_type =
+        typename Kokkos::RangePolicy<ExecParameters...>::index_type;
+
+    using neighbor_list_traits = NeighborList<NeighborListType>;
+
+    using memory_space = typename neighbor_list_traits::memory_space;
+
+    static_assert( is_accessible_from<memory_space, execution_space>{}, "" );
+
+    auto neigh_reduce = KOKKOS_LAMBDA( const index_type i, ReduceType &ival )
+    {
+        const index_type nn = neighbor_list_traits::numNeighbor( list, i );
+
+        for ( index_type n = 0; n < nn; ++n )
+        {
+            const index_type j =
+                neighbor_list_traits::getNeighbor( list, i, n );
+
+            for ( index_type a = n + 1; a < nn; ++a )
+            {
+                const index_type k =
+                    neighbor_list_traits::getNeighbor( list, i, a );
+                Impl::functorTagDispatch<work_tag>( functor, i, j, k, ival );
+            }
+        }
+    };
+    if ( str.empty() )
+        Kokkos::parallel_reduce( exec_policy, neigh_reduce, reduce_val );
```

Comment:
This seems to reduce a single value over all particles over all neighbors correct? I guess I though the MD use case was a per particle reduction over all neighbors. Hence if there are `N` particles you would get `N` reductions where the reduction is a force or something. What you have will almost work in that case - just replace the `parallel_reduce` with a `parallel_for` and have the user input a view/slice of values for each particle into which to write the reductions. Also note that the reducers can be multidimensional so you could calculate a force in 3 space dimensions or something in this way.

Reply:
OK after talking with Sam this functionality is needed for total reductions over neighbor operations (e.g. computing total energy in an MD calculculations). We will do what I described here in another PR which does something like `neighbor_parallel_for_reduce` or similar name with a custom result functor to allow the user to extract the particle-wise reductions.